crates/codegen/runtime/generator/src/parser/grammar/resolver.rs

```diff
@@ -198,55 +105,129 @@ impl ParserThunk {
-fn resolve_grammar_element(ident: &Identifier, ctx: &mut ResolveCtx<'_>) -> GrammarElement {
-    let (lex_ctx, elem) = ctx.items.get(ident).expect("Missing item");
+pub struct Resolution {
+    /// Original items as defined by the DSL v2.
+    items: HashMap<Identifier, (Identifier, Item)>,
+    resolved: HashMap<Identifier, GrammarElement>,
+}
 
-    let lex_ctx = lex_ctx
-        .clone()
-        .unwrap_or_else(|| Identifier::from("Default"));
+impl ResolveCtx {
+    pub fn resolve(lang: &Language) -> Resolution {
+        // Collect language items into a lookup table to speed up resolution
+        let items: HashMap<_, _> = lang
+            .topics()
+            .flat_map(|topic| {
+                topic.items.iter().map(|item| {
+                    let lex_ctxt = topic.lexical_context.as_ref().unwrap_or(&DEFAULT_LEX_CTXT);
 
-    // The nonterminals are mutually recursive (so will be the resolution of their definitions),
-    // so make sure to insert a thunk for nonterminals to resolve to break the cycle.
-    let inserted_thunk = match (elem, ctx.resolved.contains_key(ident)) {
-        (
-            Item::Struct { .. }
-            | Item::Enum { .. }
-            | Item::Repeated { .. }
-            | Item::Separated { .. },
-            false,
-        ) => {
-            let thunk = Rc::new(NamedParserThunk {
-                name: ident.clone(),
-                context: lex_ctx.clone(),
-                is_inline: false,
-                def: OnceCell::new(),
-            });
+                    (item.name().clone(), (lex_ctxt.clone(), item.clone()))
+                })
+            })
+            .collect();
+
+        let mut ctx = ResolveCtx {
+            items,
+            resolved: HashMap::new(),
+        };
+
+        for item in lang.items() {
+            resolve_grammar_element(item.name(), &mut ctx);
+        }
+
+        // Trivia is defined separately from the main grammar
+        let leading_trivia = Rc::new(NamedTriviaParser {
+            name: Identifier::from("LeadingTrivia"),
+            def: resolve_trivia(lang.leading_trivia.clone(), TriviaKind::Leading, &mut ctx),
+        });
+
+        let trailing_trivia = Rc::new(NamedTriviaParser {
+            name: Identifier::from("TrailingTrivia"),
+            def: resolve_trivia(lang.trailing_trivia.clone(), TriviaKind::Trailing, &mut ctx),
+        });
+
+        for trivia in [leading_trivia, trailing_trivia] {
             ctx.resolved.insert(
-                ident.clone(),
-                (Rc::clone(&thunk) as Rc<dyn ParserDefinition>).into(),
+                trivia.name().clone(),
+                GrammarElement::TriviaParserDefinition(trivia),
             );
-            Some(ParserThunk::Regular(thunk))
         }
+
+        Resolution {
+            items: ctx.items,
+            resolved: ctx.resolved,
+        }
+    }
+}
+
+impl Resolution {
+    pub fn original(&self, name: &Identifier) -> &(Identifier, Item) {
```

Comment:
Ah, I forgot that I still had it here. I need it to extract the lexical context for a given item.

I can't strictly replace it with a "get lexical context for an item" because some item may be reachable from different lexical contexts (i.e. pragma keyword) as required by our current lexer codegen/setup.

I'm happy with renaming it but are we fine with keeping this function for now?

Reply:
Sounds good. I suggest renaming it to its true purpose, and add this reasoning as a commment for the future.

Reply:
Done in https://github.com/NomicFoundation/slang/pull/1018/commits/234c9240d92750070a00761618b6a382f72e1f85